src/ServiceManagement/RecoveryServices/Commands.RecoveryServices/Service/InvokeAzureRecoveryServicesVaultUpgrade.cs

```diff
@@ -0,0 +1,238 @@
+        [ValidateNotNullOrEmpty]
+        [Alias("TargetRG", "TargetRGName", "RG")]
+        public string TargetResourceGroupName { get; set; }
+
+        /// <summary>
+        /// Gets or sets switch parameter. On passing, command does not ask for confirmation.
+        /// </summary>
+        [Parameter(Mandatory = false)]
+        public SwitchParameter Force { get; set; }
+
+        #endregion Parameters
+
+        /// <summary>
+        /// ProcessRecord of the command.
+        /// </summary>
+        public override void ExecuteCmdlet()
+        {
+            CheckVaultUpgradePrerequisitesResponse result =
+                this.ValidateVaultUpgradePrerequisites();
+
+            bool actionConfirmed = false;
+            ResourceUpgradeDetails response = null;
+            string message = result == CheckVaultUpgradePrerequisitesResponse.Succeeded ?
+                Properties.Resources.PrerequisitesCheckPassed :
+                Properties.Resources.PrerequisitesCheckReturnedWarnings;
+
+            try
+            {
+                this.ConfirmAction(
+                    this.Force.IsPresent,
+                    string.Format(message),
+                    Properties.Resources.StartVaultUpgradeWhatIfMessage,
+                    this.VaultName,
+                    () =>
+                        {
+                            response = this.RecoveryServicesClient.StartVaultUpgrade(
+                                this.VaultName,
+                                this.Location,
+                                this.ResourceType,
+                                this.TargetResourceGroupName,
+                                this.Profile.Context.Subscription.Id.ToString());
+
+                            actionConfirmed = true;
+                        });
+
+                if (actionConfirmed)
+                {
+                    this.WriteObject(new ASRVaultUpgradeDetails(response));
+                    this.WaitForJobCompletion();
+                }
+            }
+            catch (Exception exception)
+            {
+                this.RecoveryServicesClient.HandleVaultUpgradeException(exception);
+            }
+        }
+
+        /// <summary>
+        /// Track resource upgrade operation.
+        /// </summary>
+        private void WaitForJobCompletion()
+        {
+            TrackResourceUpgradeResponse response = null;
+            DateTime startTime = DateTime.Now;
+            double taskTimeoutInSeconds = 6000;
+            double elapsedSeconds = 0;
+            ProgressRecord record =
+                new ProgressRecord(
+                    0,
+                    Properties.Resources.VaultUpgradeInProgress,
+                    Properties.Resources.WaitingForCompletion);
+
+            this.WriteResponse(Properties.Resources.VaultUpgradeInProgress);
+
+            do
+            {
+                Thread.Sleep(PSRecoveryServicesClient.TimeToSleepBeforeFetchingJobDetailsAgain);
```

Comment:
Please use this method: https://github.com/Azure/azure-powershell/blob/dev/src/Common/Commands.Common/TestMockSupport.cs#L38  instead of calling Thread.Sleep directly.  This will keep mocked tests that run this cmdlet from having to wait.